### PR TITLE
Add TypeScript definition

### PR DESCRIPTION
**Pull request by BendingBender**

### CODE REVIEW

File: index.d.ts [hunk 1]
```diff
@@ -0,0 +1,287 @@
...
+	 * Stop and clear the spinner.
+	 *
+	 * @returns The spinner instance.
+	 */
+	stop(): Ora;
+
+	/**
+	 * Stop the spinner, change it to a green `✔` and persist the current text, or `text` if provided.
+	 *
+	 * @param text will persist text if provided
```
Comment (sindresorhus) [inline index.d.ts]:
```suggestion
	 * @param text - Will persist text if provided.
```

File: index.d.ts [hunk 2]
```diff
@@ -0,0 +1,287 @@
...
+	 *     interval: 80, // optional
+	 *     frames: ['-', '+', '-']
+	 * }
+	 *
+	 * @default 'dots'
+	 */
+	spinner?: SpinnerName | Spinner;
+
+	/**
+	 * Color of the spinner.
```
Comment (sindresorhus) [inline index.d.ts]:
```suggestion
	 * Color of the spinner.
	 *
```

File: index.d.ts [hunk 3]
```diff
@@ -0,0 +1,287 @@
...
+	 */
+	interval?: number;
+
+	/**
+	 * Stream to write the output.
+	 *
+	 * You could for example set this to `process.stdout` instead.
+	 * @default process.stderr
+	 */
+	stream?: NodeJS.WritableStream;
```
Comment (sindresorhus) [inline index.d.ts]: Import the stream interface instead of using the global.

File: index.d.ts [hunk 4]
```diff
@@ -0,0 +1,287 @@
...
+	 */
+	text?: string;
+}
+
+export interface Spinner {
+	interval?: number;
+	frames: string[];
+}
+
+export type SpinnerName =
```
Comment (sindresorhus) [inline index.d.ts]: This should be defined in https://github.com/sindresorhus/cli-spinners and imported from there instead.

File: index.d.ts [hunk 5]
```diff
@@ -0,0 +1,287 @@
...
+	render(): Ora;
+
+	/**
+	 * Get a new frame.
+	 * @returns The spinner instance.
+	 */
+	frame(): Ora;
+}
+
+export interface Options {
```
Comment (sindresorhus) [inline index.d.ts]: This should be defined before it's used.

File: index.d.ts [hunk 6]
```diff
@@ -0,0 +1,287 @@
...
+	 * Indent the spinner with the given number of spaces.
+	 * @default 0
+	 */
+	indent?: number;
+
+	/**
+	 * Interval between each frame.
+	 *
+	 * Spinners provide their own recommended interval, so you don't really need to specify this.
+	 * @default Provided by the spinner or 100
```
Comment (sindresorhus) [inline index.d.ts]: I think `@default` is meant for a an actual value, not a description. Can you double-check the JSDoc/TSDoc docs?

File: index.d.ts [hunk 7]
```diff
@@ -0,0 +1,287 @@
...
+	 * Set to `false` to stop Ora from hiding the cursor.
+	 * @default true
+	 */
+	hideCursor?: boolean;
+
+	/**
+	 * Indent the spinner with the given number of spaces.
+	 * @default 0
+	 */
+	indent?: number;
```
Comment (sindresorhus) [inline index.d.ts]: Make all of these `readonly`. See: https://github.com/sindresorhus/typescript-definition-style-guide#prefer-read-only-values

File: index.d.ts [hunk 8]
```diff
@@ -0,0 +1,287 @@
...
+
+/**
+ * Starts a spinner for a promise. The spinner is stopped with `.succeed()` if the promise fulfills
+ * or with `.fail()` if it rejects.
+ *
+ * @param action
+ * @param options If a string is provided, it is treated as a shortcut for `options.text`.
+ * @returns The spinner instance.
+ */
+export function promise(action: PromiseLike<any>, options?: Options | string): Ora;
```
Comment (sindresorhus) [inline index.d.ts]: Use `unknown`.

File: index.d.ts [hunk 9]
```diff
@@ -0,0 +1,287 @@
...
+ *
+ * @param options If a string is provided, it is treated as a shortcut for `options.text`.
+ */
+export default function ora(options?: Options | string): Ora;
+
+/**
+ * Starts a spinner for a promise. The spinner is stopped with `.succeed()` if the promise fulfills
+ * or with `.fail()` if it rejects.
+ *
+ * @param action
```
Comment (sindresorhus) [inline index.d.ts]: Should have a description or be left out.

File: index.d.ts [hunk 10]
```diff
@@ -0,0 +1,287 @@
...
+	 *
+	 * @param text will persist text if provided
+	 * @returns The spinner instance.
+	 */
+	info(text?: string): Ora;
+
+	/**
+	 * Stop the spinner and change the symbol or text.
+	 *
+	 * @param options
```
Comment (sindresorhus) [inline index.d.ts]: > If the parameter is options it doesn't need a description.
> \- https://github.com/sindresorhus/typescript-definition-style-guide#prefer-read-only-values

File: index.d.ts [hunk 11]
```diff
@@ -0,0 +1,287 @@
...
+export interface Options {
+	/**
+	 * Text to display after the spinner.
+	 */
+	text?: string;
+
+	/**
+	 * Name of one of the provided spinners. See [`example.js`](https://github.com/BendingBender/ora/blob/master/example.js)
+	 * in this repo if you want to test out different spinners. On Windows, it will always use the line spinner as
+	 * the Windows command-line doesn't have proper Unicode support.
```
Comment (sindresorhus) [inline index.d.ts]: Don't hard-wrap.

File: index.d.ts [hunk 12]
```diff
@@ -0,0 +1,287 @@
...
+	 *
+	 * Or an object like:
+	 *
+	 * @example
```
Comment (sindresorhus) [inline index.d.ts]: Example should come last.

File: index.test-d.ts [hunk 13]
```diff
@@ -0,0 +1,48 @@
...
+ora({ text: 'Loading unicorns' });
+ora({ spinner: 'squish' });
+ora({ spinner: { frames: ['-', '+', '-'] } });
+ora({ spinner: { interval: 80, frames: ['-', '+', '-'] } });
+ora({ color: 'cyan' });
+ora({ hideCursor: true });
+ora({ indent: 1 });
+ora({ interval: 80 });
+ora({ stream: new PassThrough() });
+ora({ isEnabled: true });
```
Comment (sindresorhus) [inline index.test-d.ts]:
```suggestion
ora({isEnabled: true});
```

For all braces.

File: package.json [hunk 14]
```diff
@@ -35,6 +36,7 @@
 		"idle"
 	],
 	"dependencies": {
+		"@types/node": "^11.9.5",
```
Comment (sindresorhus) [inline package.json]: Should this really be a dependency? I think it's expected that Node.js project import this themselves.

File: index.d.ts (hunk 8 above)
Comment (BendingBender) [reply index.d.ts]: Not relevant here and better left as `any`. This interface accepts any input, not some input. `unknown` would be relevant if we would return this value to the user.

File: index.d.ts (hunk 3 above)
Comment (BendingBender) [reply index.d.ts]: Not possible here, NodeJS is a global namespace, not a module you can import.

File: index.d.ts (hunk 4 above)
Comment (BendingBender) [reply index.d.ts]: Ok, will do the types for cli-spinners first.

File: index.test-d.ts (hunk 13 above)
Comment (BendingBender) [reply index.test-d.ts]: Is it allowed to add a Prettier config to this project? My default settings conflict with this and I keep forgetting editing this back after pretty-printing.

File: package.json (hunk 14 above)
Comment (BendingBender) [reply package.json]: Seems to be the case: https://www.typescriptlang.org/docs/handbook/declaration-files/publishing.html

File: index.d.ts (hunk 3 above)
Comment (sindresorhus) [reply index.d.ts]: It is possible: https://github.com/sindresorhus/hook-std/blob/f429639e9d60ae3259a17cd4aefaf62a9e4f15c1/index.d.ts#L1

File: index.test-d.ts (hunk 13 above)
Comment (sindresorhus) [reply index.test-d.ts]: No, you should not have Prettier enabled when there's no prettier config in a project.

File: package.json (hunk 14 above)
Comment (sindresorhus) [reply package.json]: That only talks about npm packages, not built-in Node.js modules. I've been putting it in devDependencies in other packages, and it's fine: https://github.com/sindresorhus/hook-std/blob/f429639e9d60ae3259a17cd4aefaf62a9e4f15c1/package.json#L40

File: index.d.ts (hunk 8 above)
Comment (sindresorhus) [reply index.d.ts]: `unknown` works exactly the same as `any` here. I prefer using `unknown` for consistency.

**Comment by sindresorhus** [comment]
Can you not force push? It makes it impossible to see what changed. I will squash on merge anyway ;)

File: index.d.ts (hunk 8 above)
Comment (BendingBender) [reply index.d.ts]: What do you think about this: https://github.com/sindresorhus/p-event/pull/21#discussion_r260744213 ?

**Pull request merged**